From 73ded2e64e1a008df5cd46358d59c8421c0d6989 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Wed, 15 Nov 2023 12:57:06 +0100
Subject: [PATCH] PCI: stm32: set bus_dma_limit on PCIe devices

DMA cannot access the last 2GB above the 4GB addressing space
when the DDR size is > 4GB. Limit the bus_dma_limit to 32 bit.

Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: If4015d6c31ed28edcbc467d623504f3ff26b0c8c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/339957
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/pci/controller/dwc/pcie-stm32.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

--- a/drivers/pci/controller/dwc/pcie-stm32.c
+++ b/drivers/pci/controller/dwc/pcie-stm32.c
@@ -406,6 +406,26 @@ static void quirk_stm32_pcie_limit_mrrs(
 DECLARE_PCI_FIXUP_HEADER(PCI_ANY_ID, PCI_ANY_ID,
 			 quirk_stm32_pcie_limit_mrrs);
 
+static int stm32_dma_limit(struct pci_dev *pdev, void *data)
+{
+	dev_dbg(&pdev->dev, "set bus_dma_limit");
+
+	pdev->dev.bus_dma_limit = DMA_BIT_MASK(32);
+
+	return 0;
+}
+
+static void quirk_stm32_dma_mask(struct pci_dev *pci)
+{
+	struct pci_dev *root_port;
+
+	root_port = pcie_find_root_port(pci);
+
+	if (root_port && is_stm32_pcie_driver(root_port->dev.parent))
+		pci_walk_bus(pci->bus, stm32_dma_limit, NULL);
+}
+DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_SYNOPSYS, 0x0550, quirk_stm32_dma_mask);
+
 module_platform_driver(stm32_pcie_driver);
 MODULE_DESCRIPTION("STM32MP25 PCIe Controller driver");
 MODULE_LICENSE("GPL");
